### ISSUE TITLE
DeprecationWarning: pkg_resources is deprecated as an API

### ISSUE THREAD

**Issue by cottsay** [colcon/colcon-core#552]

**Comment by stratakis** [comment]
The deprecation warning leaks to other packages down the stack though which utilize colcon-core.

More specifically by testing the latest setuptools version on Fedora, colcon-ros and colcon-notification fails due to colcon-core's pkg_resources import.

**Comment by cottsay** [comment]
I've been toying with solutions here.

One challenge is that we have functions in `colcon-core` public API which return `pkg_resources` classes. This was probably not a great design decision. The corresponding classes in `importlib.metadata` are not similar enough to be drop-in replacements, so I think we'll need to introduce some sort of wrapper class that looks like the `pkg_resources` classes.

**Comment by mikepurvis** [comment]
Thanks for beginning to think about a way forward here—we're hitting these warnings when testing colcon plugins on Python 3.10 (in Nix). This is how to silence the warnings in pytest, but unfortunately you can't specify the actual package doing the importing, so it's a bit of a blunt instrument:

```
[tool.pytest.ini_options]
filterwarnings = [
  # This is transitively imported by colcon-core, and won't be dropped until that package
  # no longer supports Python 3.8 (and therefore Ubuntu Focal).
  "ignore::DeprecationWarning:pkg_resources",
]```

 A further complexifier is that the replacement functionality in `importlib` isn't available in Python 3.8, so either the old implementation must also be retained as long as Ubuntu Focal is a target, or else a dependency must be introduced on a pypi package supplying backports of the new functionality, eg:

- https://pypi.org/project/importlib-metadata/
- https://pypi.org/project/importlib-resources/

**Comment by cottsay** [comment]
> a dependency must be introduced on a pypi package supplying backports of the new functionality

It seems that `python3-importlib-metadata` is [available for Focal as a deb](https://packages.ubuntu.com/focal/python3-importlib-metadata). It is my intent to declare a dependency on that package conditionally on the interpreter version.

As far as I can tell, it does mean dropping debs for Bionic which reached EOL this past May.
